bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -3444,84 +3444,119 @@ private static void convertMapToStruct(WorkerExecutionContext ctx, int[] operand
+        BMap<String, BValue> bMap = (BMap<String, BValue>) sf.refRegs[i];
+
+        try {
+            sf.refRegs[j] = convertMapToStruct(ctx, bMap, structType);
+        } catch (BallerinaException e) {
+            sf.refRegs[j] = null;
+            String errorMsg = "cannot convert '" + bMap.getType() + "' to type '" + structType + ": " +
+                    e.getMessage();
+            handleTypeConversionError(ctx, sf, j, errorMsg);
+        }
+    }
+
+    private static BMap<String, BValue> convertMapToStruct(WorkerExecutionContext ctx,
+                                                           BMap<String, BValue> bMap, BStructureType structType) {
         BMap<String, BValue> bStruct = new BMap<>(structType);
-        StructureTypeInfo structInfo = ctx.callableUnitInfo
-                .getPackageInfo().getStructInfo(structType.getName());
+        StructureTypeInfo structInfo = ctx.callableUnitInfo.getPackageInfo().getStructInfo(structType.getName());
 
         for (StructFieldInfo fieldInfo : structInfo.getFieldInfoEntries()) {
             String key = fieldInfo.getName();
             BType fieldType = fieldInfo.getFieldType();
             BValue mapVal = null;
-            try {
-                boolean containsField = bMap.hasKey(key);
-                DefaultValueAttributeInfo defaultValAttrInfo = null;
-                if (containsField) {
-                    mapVal = bMap.get(key);
-                    if (mapVal == null && BTypes.isValueType(fieldType)) {
-                        throw BLangExceptionHelper.getRuntimeException(
-                                RuntimeErrors.INCOMPATIBLE_FIELD_TYPE_FOR_CASTING, key, fieldType, null);
+
+            boolean containsField = bMap.hasKey(key);
+            DefaultValueAttributeInfo defaultValAttrInfo = null;
+            if (containsField) {
+                mapVal = bMap.get(key);
+                if (mapVal == null && BTypes.isValueType(fieldType)) {
+                    throw BLangExceptionHelper.getRuntimeException(
+                            RuntimeErrors.INCOMPATIBLE_FIELD_TYPE_FOR_CASTING, key, fieldType, null);
+                }
+
+                if (mapVal != null) {
+                    BType mapValType = mapVal.getType();
+                    if (mapValType.getTag() == TypeTags.MAP_TAG) {
+                        if (fieldType.getTag() == TypeTags.RECORD_TYPE_TAG) {
+                            bStruct.put(key, convertMapToStruct(ctx, (BMap<String, BValue>) mapVal,
+                                                                ((BStructureType) fieldType)));
+                            continue;
+                        } else if (fieldType.getTag() == TypeTags.UNION_TAG) {
+                            boolean conversionDone = false;
+                            for (BType memType : ((BUnionType) fieldType).getMemberTypes()) {
+                                if (memType.getTag() == TypeTags.RECORD_TYPE_TAG) {
```

Comment:
There are 7 nested blocks. Is it possible to reduce the number of nested blocks?